quinn-proto/src/connection.rs

```diff
@@ -1961,173 +1961,197 @@ impl Connection {
+            let partial_encode = header.encode(&mut buf);
+            coalesce = coalesce && !header.is_short();
+
+            if probe && ack_only && header.is_1rtt() {
+                // Nothing ack-eliciting to send, so we need to make something up
+                self.ping_pending = true;
+            }
+            ack_only &= !self.ping_pending;
+
+            let sent = if close {
+                trace!(self.log, "sending CONNECTION_CLOSE");
+                let max_len = buf.capacity()
+                    - partial_encode.start
+                    - partial_encode.header_len
+                    - space.crypto.as_ref().unwrap().packet.tag_len();
+                match self.state {
+                    State::Closed(state::Closed {
+                        reason: state::CloseReason::Application(ref x),
+                    }) => x.encode(&mut buf, max_len),
+                    State::Closed(state::Closed {
+                        reason: state::CloseReason::Connection(ref x),
+                    }) => x.encode(&mut buf, max_len),
+                    _ => unreachable!(
+                        "tried to make a close packet when the connection wasn't closed"
+                    ),
+                }
+                coalesce = false;
+                None
+            } else if let Some((path_remote, token)) = self.offpath_responses.pop() {
+                // For simplicity's sake, we don't bother trying to batch together or deduplicate path
+                // validation probes.
+                trace!(self.log, "PATH_RESPONSE {token:08x}", token = token);
+                buf.write(frame::Type::PATH_RESPONSE);
+                buf.write(token);
+                remote = path_remote;
+                coalesce = false;
+                None
+            } else {
+                Some(self.populate_packet(now, space_id, &mut buf))
+            };
 
-        let space = &mut self.spaces[space_id as usize];
-        let crypto = if let Some(ref crypto) = space.crypto {
-            crypto
-        } else if space_id == SpaceId::Data {
-            self.zero_rtt_crypto.as_ref().unwrap()
-        } else {
-            unreachable!("tried to send {:?} packet without keys", space_id);
-        };
+            let space = &mut self.spaces[space_id as usize];
+            let crypto = if let Some(ref crypto) = space.crypto {
+                crypto
+            } else if space_id == SpaceId::Data {
+                self.zero_rtt_crypto.as_ref().unwrap()
+            } else {
+                unreachable!("tried to send {:?} packet without keys", space_id);
+            };
 
-        let mut padded = if self.side.is_client() && space_id == SpaceId::Initial {
-            // Initial-only packets MUST be padded
-            buf.resize(MIN_INITIAL_SIZE - crypto.packet.tag_len(), 0);
-            true
-        } else {
-            false
-        };
+            let mut padded = if exact_number == 0 && self.side.is_client() &&
```

Comment:
The client's first flight might be retransmitted and retransmits do not reuse packet numbers. Initial packets may also be sent after the client's first flight but before handshake keys are available (e.g. ACKs), and those too must be padded.

Reply:
Can you take a look at the current approach? Seems nice, though I haven't fully thought through the interactions with `close` and `offpath_responses` -- not sure how likely these are to occur at this stage?

Reply:
Close and, I think, challenge responses can both be sent during the handshake, but I think the interactions here are correct.